collector/src/etw_parser.rs

```diff
@@ -0,0 +1,677 @@
+    - Validate every Pmc has a next CSwitch event at the same timestamp
+- It may be convenient at this point to rewrite into a stream of (pmc data, CSwitch data)
+- Count up deltas of PMC data on each CPU -> this is the total for the counter
+- Return results
+*/
+
+#[derive(Debug, Eq, PartialEq)]
+struct EventHeader {
+    name: String,
+    columns: Vec<String>,
+}
+
+const PROCESS_START: &str = "P-Start";
+const PROCESS_END: &str = "P-End";
+const PMC: &str = "Pmc";
+const CSWITCH: &str = "CSwitch";
+const TIMESTAMP: &str = "TimeStamp";
+const PROCESS_NAME_PID: &str = "Process Name ( PID)";
+const THREAD_ID: &str = "ThreadID";
+const INSTRUCTION_RETIRED: &str = "InstructionRetired";
+const TOTAL_CYCLES: &str = "TotalCycles";
+const NEW_PROCESS_NAME_PID: &str = "New Process Name ( PID)";
+const OLD_PROCESS_NAME_PID: &str = "Old Process Name ( PID)";
+const CPU: &str = "CPU";
+const PARENT_PID: &str = "ParentPID";
+
+fn parse_header(r: &mut dyn BufRead) -> anyhow::Result<Vec<EventHeader>> {
+    let mut line = String::new();
+    r.read_line(&mut line)?;
+
+    anyhow::ensure!(line.trim() == "BeginHeader", "BeginHeader line not found at beginning of file");
+
+    let mut headers = Vec::with_capacity(6);
+
+    while { line.clear(); r.read_line(&mut line)? != 0 } {
+        let l = line.trim();
+        if l == "EndHeader" {
+            break;
+        }
+
+        let mut contents = l.split(',');
+        let name = contents.next().expect("event header was missing name").trim();
+
+        if name == PROCESS_START || name == PROCESS_END || name == PMC || name == CSWITCH {
+            headers.push(EventHeader {
+                name: name.to_string(),
+                columns: contents.map(|s| s.trim().to_string()).collect(),
+            });
+        }
+    }
+
+    anyhow::ensure!(line.trim() == "EndHeader", "EndHeader line not found");
+
+    Ok(headers)
+}
+
+fn validate_os_header_line(r: &mut dyn BufRead) -> anyhow::Result<()> {
+    let mut line = String::new();
+    r.read_line(&mut line)?;
+
+    anyhow::ensure!(line.starts_with("OS Version"), "OS version line not found");
+
+    let components: Vec<_> = line.split(',').collect();
+
+    let get_value = |key: &str| {
+        components
+            .iter()
+            .find(|c| c.trim_start().starts_with(key))
+            .expect("key not found")
+            .split(':')
+            .nth(1)
+            .unwrap()
+            .trim()
+    };
+
+    let events_lost: u32 = get_value("Events Lost:").parse()?;
+    anyhow::ensure!(events_lost == 0, "events were lost");
+
+    let buffers_lost: u32 = get_value("Buffers lost:").parse()?;
+    anyhow::ensure!(buffers_lost == 0, "buffers were lost");
+
+    Ok(())
+}
+
+#[derive(Debug, Eq, PartialEq)]
+struct Pmc {
+    timestamp: u64,
+    thread_id: u64,
+    instructions_retired: u64,
+    total_cycles: u64,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+struct CSwitch {
```

Comment:
I assume this is a context switch? It's fine to use the short form if that's what's normally used, I'd just note as a doc.